### ISSUE TITLE
RPCs causing clients to disconnect from server

### ISSUE THREAD

**Issue by ForgeUser2018** [BeardedManStudios/ForgeNetworkingRemastered#201]
### Version Number and Operating System(s):
- Cloned master repo on 2018-10-20
- Fresh Unity project
- Windows 10 64bit
- Unity 2018.2.11f1

### Expected behavior:
Sending RPCs should be stable, both reliable and unreliable, regardless of the bandwidth received and sent by both the clients and server.

### Actual behavior:
RPCs do work initially, but no matter how RPCs are sent, e.g. by using `SendRPC()` or `SendRPCUnreliable()`, after a certain amount of data has been transmitted (in my tests ~1.1MB), clients get disconnected and sometimes crash or become unresponsive. After a client reconnects, the procedure starts over. 

However, the issue only occurs when more than one client is connected at a time. In every test, one client remained active. Example: 5 clients + 1 server running simultaneously, as soon as the issue kicks in, 4 of the 5 clients get disconnected and 1 keeps running fine even after e.g. 1 hour. During the tests the server itself never crashed.

I've also tested RPC calls with different receivers (All, Others, OthersBuffered (replacing previous)), same issue. I lowered the send rate, but this only made the test run longer until the clients got disconnected (from 17 minutes to about 2 hours), so the cause of the problem is likely related to the amount of data sent across the network through RPCs. I also sent empty RPCs and left the receiving callback empty as well, but to no avail, same results. I hooked into the `playerTimeout` event and also raised the `TimeoutMilliseconds`, but the timeout event was never called. No matter if the clients run in the background or are actively used, on every test the clients got disconnected, no matter what. Furthermore, it doesn't make a difference when only one client is sending data or all simultaneously.

According to the profiler, memory usage and GC allocation are negligible. `Run in Background` is ticked, so is `Use Main Thread Manager`. Tests utilizing Binary Frame directly and NCW fields ran fine. As soon as RPCs come into play, the issue comes up.

### Steps to reproduce:
Open the vanilla Cube Forge example scene. Inside the NCW, open the `NetworkCameraNetworkObject`, add a `position` RPC and save. Open the `NetCam.cs` script, exchange the position updates `networkObject.position` with RPCs and add a callback handling the position updates when data is received. Edit `NetworkStart()` and `Update()` to adapt to the changes made. Build the game, run the host inside Unity and start and connect 5 clients. Either keep moving manually (e.g. keep pressing W) or edit the script to randomly move around for you. After a certain amount of data has been sent across the network, clients start to disconnect.

> **_So to reproduce this issue, you basically need to keep sending data across the network through a RPC. The payload and send interval do not matter, as the issue is related to the total amount of data transmitted._**

**EDIT:**
Check out my Unity Project for simplified reproduction of this issue:
https://github.com/ForgeUser2018/Forge-Networking-Remastered-RPC-Issue-Test

#### [Optional] Discord Username:
ForgeUser2018

**Comment by nanospork** [comment]
I'm running into this too, in my active project. Hosting in one instance, client connected in the other. After a few minutes of constant RPCs, the client disconnects but the host is fine. Nothing is showing up in bmslog.txt, even though I have `Log to File` enabled.

I don't know much about how Forge's networking works internally yet (still very new), but it smells like a full buffer or something?

**Comment by nanospork** [comment]
I'm actually getting crashes on the host now as well. Still seems to be related to the amount of data sent over RPC. The crashed host is exhibiting the same CPU consumption increase when it freezes.

**Comment by ForgeUser2018** [comment]
@nanospork Yea looks like there is a buffer overflow. Maybe there is a deserialization problem somewhere. Could also be a stack overflow caused by a faulty (while-) loop or alike. Dunno, would be helpful if the FNR dev would take a peek at this.

**Comment by ForgeUser2018** [comment]
Commenting out lines `394` till `399` in file [`UDPServer.cs`](https://github.com/BeardedManStudios/ForgeNetworkingRemastered/blob/f1ccd8704e75f38f49b70f9fe0f1867739427bff/BeardedManStudios/Source/Forge/Networking/UDPServer.cs#L394-L399) prevents the clients from being disconnected. So this part needs to be debugged.

**Comment by Rafaesp** [comment]
Very serious bug. I hope it gets fixed soon.
Thanks @ForgeUser2018 for all the work done.

**Comment by AttackingHobo** [comment]
How is the status of this bug. I was thinking about using Forge, but this seems to be a killer bug right now.

**Comment by ForgeUser2018** [comment]
@AttackingHobo No fix in sight. This bug renders FNR pretty much unusable.

**Comment by NFMynster** [comment]
@BrentFarris

**Comment by BrentFarris** [comment]
Thanks for the report, @Crazy8ball and myself will look into the issue, a lot of the times these kinds of things are usually a minor fix so hopefully, it can be identified quickly. If anyone else has debugged the code and found the spot where this would mainly be happening, please let us know. @ForgeUser2018 I'll take a look around the area of code you were mentioning in your comment as well.

Thanks for the report!

**Comment by BrentFarris** [comment]
Please review this pull request:  https://github.com/BeardedManStudios/ForgeNetworkingRemastered/commit/ff31f675f553a326ad2be04d21a97b894ecea5fd

**TLDR;**
It will resolve the issue that is found here, I'll be creating a new issue to find out where the garbage "connect" message is being relayed from on the server that is causing this issue.

**Long explanation:**
What was found was that the server is sending a garbage message to the client that looks like a connect message that a client would send to a server. The client is then saying that it has read the message (but did nothing with it). This is a reliable message so it was telling the server that it has successfully collected the message from the network. This message to the server saying it has successfully collected the connect message was identified by the server as the client trying to connect **again** to the server, which is not allowed and triggered the disconnect.

There will be another issue opened to investigate the garbage message sent from the server and identify why / who is sending it from the list of clients.
